### PR TITLE
Maximum bipartite matching algorithm

### PR DESCRIPTION
**Pull request by jmatak**

### CODE REVIEW

File: cpp/bipartite_matching_module/bipartite_matching_test.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+TEST(BipartiteMatching, Performance) {
+  auto seed = 42;
+  std::mt19937 rng(seed);
+
+  auto num_of_nodes = 250;
+  auto num_of_edges = num_of_nodes * num_of_nodes / 5;
+
+  std::set<std::pair<uint64_t, uint64_t>> edges;
+  for (int i = 0; i < num_of_edges; ++i) {
+    std::uniform_int_distribution<> dist(1, num_of_nodes);
```
Comment (jbajic) [inline cpp/bipartite_matching_module/bipartite_matching_test.cpp:130]: Move the dist out of loop

File: cpp/bipartite_matching_module/bipartite_matching_test.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    auto from = dist(rng), to = dist(rng);
+    edges.insert({from, to});
+  }
+
+  mg_test_utility::Timer timer;
+  bipartite_matching_util::MaximumMatching(std::vector<std::pair<uint64_t, uint64_t>>(edges.begin(), edges.end()));
+  auto time_elapsed = timer.Elapsed();
+  ASSERT_TRUE(timer.Elapsed() < std::chrono::seconds(1));
```
Comment (jbajic) [inline cpp/bipartite_matching_module/bipartite_matching_test.cpp:138]:
```suggestion
  auto time_elapsed = timer.Elapsed();
  ASSERT_TRUE(time_elapsed < std::chrono::seconds(1));
```

File: cpp/bipartite_matching_module/bipartite_matching_test.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+TEST(BipartiteMatching, Performance) {
+  auto seed = 42;
+  std::mt19937 rng(seed);
+
+  auto num_of_nodes = 250;
+  auto num_of_edges = num_of_nodes * num_of_nodes / 5;
+
+  std::uniform_int_distribution<> dist(1, num_of_nodes);
+  std::set<std::pair<uint64_t, uint64_t>> edges;
+  for (int i = 0; i < num_of_edges; ++i) {
```
Comment (jbajic) [inline cpp/bipartite_matching_module/bipartite_matching_test.cpp:130]:
```suggestion
  for (std::size_t i = 0; i < num_of_edges; ++i) {
```

File: cpp/bipartite_matching_module/algorithm/bipartite_matching.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+bool IsGraphBipartite(const mg_graph::GraphView<> &graph) {
+  auto node_size = graph.Nodes().size();
+
+  // -1 to indicate that color is not set to that node
+  std::vector<std::int8_t> colors(node_size, -1);
+
+  for (std::uint64_t i = 0; i < node_size; i++) {
+    if (colors[i] == -1) {
+      if (!IsSubgraphBipartite(graph, colors, i)) {
```
Comment (antonio2368) [inline cpp/bipartite_matching_module/algorithm/bipartite_matching.cpp:30]: You can connect this into a single if.

File: cpp/bipartite_matching_module/algorithm/bipartite_matching.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      }
+    }
+  }
+
+  return true;
+}
+
+std::uint64_t MaximumMatching(const std::vector<std::pair<std::uint64_t, std::uint64_t>> &bipartite_edges) {
+  std::set<std::uint64_t> group_a;
+  std::set<std::uint64_t> group_b;
```
Comment (antonio2368) [inline cpp/bipartite_matching_module/algorithm/bipartite_matching.cpp:69]: Is the order of the nodes important?

File: cpp/bipartite_matching_module/algorithm/bipartite_matching.cpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+namespace bipartite_matching_alg {
+
+std::uint64_t BipartiteMatching(const mg_graph::GraphView<> &graph) {
+  if (!bipartite_matching_util::IsGraphBipartite(graph)) return 0;
+
+  std::vector<std::pair<std::uint64_t, std::uint64_t>> disjoint_edges;
+
+  std::map<std::uint64_t, std::uint64_t> first_set;
+  std::map<std::uint64_t, std::uint64_t> second_set;
```
Comment (antonio2368) [inline cpp/bipartite_matching_module/algorithm/bipartite_matching.cpp:106]: Is the order of nodes important?

**Review by jbajic — APPROVED** [review]

File: cpp/bipartite_matching_module/algorithm/bipartite_matching.cpp (hunk 5 above)
Comment (jmatak) [reply cpp/bipartite_matching_module/algorithm/bipartite_matching.cpp:69]: It's not, therefore I am adding unordered_set and unordered_map, where suitable

File: cpp/bipartite_matching_module/bipartite_matching_test.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+#include "algorithm/bipartite_matching.hpp"
+
+TEST(BipartiteMatching, RandomCompleteBipartiteGraphs) {
+  auto seed = 42;
+  auto num_of_experiments = 200;
+  std::mt19937 rng(seed);
+  for (int t = 0; t < num_of_experiments; ++t) {
+    auto max_size = 250;
+    std::uniform_int_distribution<> dist(1, max_size);
```
Comment (antonio2368) [inline cpp/bipartite_matching_module/bipartite_matching_test.cpp:17]: you can take this out of for loop, it's always the same.

**Review by jbajic — APPROVED** [review]

**Pull request merged**